src/Compilers/VisualBasic/Portable/Syntax/SyntaxFactory.vb

```diff
@@ -532,5 +532,64 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return SyntaxEquivalence.AreEquivalent(oldList.Node, newList.Node, ignoreChildNode, topLevel:=False)
         End Function
 
+        ''' <summary>
+        ''' Determites if a submission is complete.
+        ''' Returns false if the syntax is valid but incomplete.
+        ''' Returns true if the syntax is invalid or complete.
+        ''' </summary>
+        ''' <param name="tree">Syntax tree.</param>
+        Public Shared Function IsCompleteSubmission(tree As SyntaxTree) As Boolean
+            Dim options As VisualBasicParseOptions = DirectCast(tree.Options, VisualBasicParseOptions)
+            Dim languageVersion As LanguageVersion = options.LanguageVersion
+
+            If (tree Is Nothing) Then
+                Throw New ArgumentNullException(NameOf(tree))
+            End If
+
+            If (Not tree.HasCompilationUnitRoot) Then
+                Return False
+            End If
+
+            Dim compilation As CompilationUnitSyntax = DirectCast(tree.GetRoot(), CompilationUnitSyntax)
+            If Not compilation.HasErrors Then
+                Return True
+            End If
+
+            For Each err In compilation.GetDiagnostics()
+                Select Case DirectCast(err.Code, ERRID)
+                    Case ERRID.ERR_LbExpectedEndIf,
+                         ERRID.ERR_ExpectedEndRegion
+                        Return False
+                    Case ERRID.ERR_ExpectedEOS
+                        Return True
+                End Select
+            Next
+
+            Dim lastNode = compilation.ChildNodes().LastOrDefault()
```

Comment:
This finds the last top level statement. And we are interested whether the last top level statement was completed.
Looking for descendants would find the last token instead.